File: negative-inventory-oversell/node/fix-negative-inventory.js
/**
 * Find and correct Shopify variants that oversold into negative inventory.
 *
 * When the inventory policy allows selling past zero (CONTINUE), a burst of orders
 * can drive a location's available count below zero. Negative stock skews reports
 * and reorder math. This lists variants, finds locations where available is negative,
 * and sets them back to zero with inventorySetQuantities using compareQuantity so a
 * concurrent change is not clobbered. Run on a schedule. Safe to run again.
 *
 * Guide: https://www.allanninal.dev/shopify/negative-inventory-oversell/
 */
import { pathToFileURL } from "node:url";

const SHOP = process.env.SHOPIFY_SHOP || "example.myshopify.com";
const TOKEN = process.env.SHOPIFY_ACCESS_TOKEN || "shpat_dummy";
const API_VERSION = process.env.SHOPIFY_API_VERSION || "2025-01";
const ENDPOINT = `https://${SHOP}/admin/api/${API_VERSION}/graphql.json`;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function availableOf(level) {
  for (const q of level.quantities || []) {
    if (q.name === "available") return q.quantity;
  }
  return null;
}

export function oversoldLevels(variant) {
  const out = [];
  const item = variant.inventoryItem || {};
  for (const lvl of item.inventoryLevels?.nodes || []) {
    const avail = availableOf(lvl);
    if (avail !== null && avail < 0) {
      out.push({ inventoryItemId: item.id, locationId: lvl.location?.id, available: avail });
    }
  }
  return out;
}

async function gql(query, variables = {}) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Shopify ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const VARIANTS_QUERY = `
query($cursor: String) {
  productVariants(first: 50, after: $cursor, query: "inventory_quantity:<0") {
    pageInfo { hasNextPage endCursor }
    nodes {
      id sku
      inventoryItem {
        id
        inventoryLevels(first: 20) {
          nodes {
            location { id name }
            quantities(names: ["available"]) { name quantity }
          }
        }
      }
    }
  }
}`;

const SET_MUTATION = `
mutation($input: InventorySetQuantitiesInput!) {
  inventorySetQuantities(input: $input) {
    inventoryAdjustmentGroup { id }
    userErrors { field message }
  }
}`;

async function* oversoldVariants() {
  let cursor = null;
  while (true) {
    const data = (await gql(VARIANTS_QUERY, { cursor })).productVariants;
    for (const node of data.nodes) yield node;
    if (!data.pageInfo.hasNextPage) return;
    cursor = data.pageInfo.endCursor;
  }
}

async function correct(inventoryItemId, locationId, current) {
  const result = (await gql(SET_MUTATION, {
    input: {
      name: "available",
      reason: "correction",
      ignoreCompareQuantity: false,
      quantities: [{ inventoryItemId, locationId, quantity: 0, compareQuantity: current }],
    },
  })).inventorySetQuantities;
  if (result.userErrors.length) throw new Error(JSON.stringify(result.userErrors));
}

export async function run() {
  let fixed = 0;
  for await (const variant of oversoldVariants()) {
    for (const lvl of oversoldLevels(variant)) {
      console.warn(`Variant ${variant.sku || variant.id} at ${lvl.locationId} is ${lvl.available}. ${DRY_RUN ? "would set to 0" : "setting to 0"}`);
      if (!DRY_RUN) await correct(lvl.inventoryItemId, lvl.locationId, lvl.available);
      fixed++;
    }
  }
  console.log(`Done. ${fixed} oversold level(s) ${DRY_RUN ? "to correct" : "corrected"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: negative-inventory-oversell/python/fix_negative_inventory.py
"""Find and correct Shopify variants that oversold into negative inventory.

When the inventory policy allows selling past zero (CONTINUE), a burst of orders can
drive a location's available count below zero. Negative stock skews reports and
reorder math. This lists variants, finds locations where available is negative, and
sets them back to zero with inventorySetQuantities using compareQuantity so a
concurrent change is not clobbered. Run on a schedule. Safe to run again and again.
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("fix_negative_inventory")

SHOP = os.environ["SHOPIFY_SHOP"]
TOKEN = os.environ["SHOPIFY_ACCESS_TOKEN"]
API_VERSION = os.environ.get("SHOPIFY_API_VERSION", "2025-01")
ENDPOINT = f"https://{SHOP}/admin/api/{API_VERSION}/graphql.json"
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

VARIANTS_QUERY = """
query($cursor: String) {
  productVariants(first: 50, after: $cursor, query: "inventory_quantity:<0") {
    pageInfo { hasNextPage endCursor }
    nodes {
      id sku
      inventoryItem {
        id
        inventoryLevels(first: 20) {
          nodes {
            location { id name }
            quantities(names: ["available"]) { name quantity }
          }
        }
      }
    }
  }
}"""

SET_MUTATION = """
mutation($input: InventorySetQuantitiesInput!) {
  inventorySetQuantities(input: $input) {
    inventoryAdjustmentGroup { id }
    userErrors { field message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        ENDPOINT,
        json={"query": query, "variables": variables or {}},
        headers={"X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def available_of(level):
    for q in level.get("quantities") or []:
        if q.get("name") == "available":
            return q.get("quantity")
    return None


def oversold_levels(variant):
    """Return the locations where available is negative, with the values to correct."""
    out = []
    item = variant.get("inventoryItem") or {}
    for lvl in ((item.get("inventoryLevels") or {}).get("nodes") or []):
        avail = available_of(lvl)
        if avail is not None and avail < 0:
            out.append({
                "inventoryItemId": item.get("id"),
                "locationId": (lvl.get("location") or {}).get("id"),
                "available": avail,
            })
    return out


def correct(inventory_item_id, location_id, current):
    result = gql(SET_MUTATION, {"input": {
        "name": "available",
        "reason": "correction",
        "ignoreCompareQuantity": False,
        "quantities": [{
            "inventoryItemId": inventory_item_id,
            "locationId": location_id,
            "quantity": 0,
            "compareQuantity": current,
        }],
    }})["inventorySetQuantities"]
    if result["userErrors"]:
        raise RuntimeError(result["userErrors"])


def oversold_variants():
    cursor = None
    while True:
        data = gql(VARIANTS_QUERY, {"cursor": cursor})["productVariants"]
        for node in data["nodes"]:
            yield node
        if not data["pageInfo"]["hasNextPage"]:
            return
        cursor = data["pageInfo"]["endCursor"]


def run():
    fixed = 0
    for variant in oversold_variants():
        for lvl in oversold_levels(variant):
            log.warning("Variant %s at %s is %s. %s", variant.get("sku") or variant["id"],
                        lvl["locationId"], lvl["available"], "would set to 0" if DRY_RUN else "setting to 0")
            if not DRY_RUN:
                correct(lvl["inventoryItemId"], lvl["locationId"], lvl["available"])
            fixed += 1
    log.info("Done. %d oversold level(s) %s.", fixed, "to correct" if DRY_RUN else "corrected")


if __name__ == "__main__":
    run()
